core/src/main/java/com/linecorp/armeria/internal/common/util/CertificateUtil.java

```diff
@@ -0,0 +1,136 @@
+
+import java.io.File;
+import java.io.InputStream;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSessionContext;
+
+import org.bouncycastle.asn1.x500.RDN;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+import org.bouncycastle.asn1.x500.style.IETFUtils;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateHolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.handler.ssl.ApplicationProtocolNegotiator;
+import io.netty.handler.ssl.SslContext;
+
+public final class CertificateUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(CertificateUtil.class);
+
+    private static final LoadingCache<X509Certificate, String> commonNameCache =
+            Caffeine.newBuilder()
+                    .weakKeys()
+                    .build(cert -> {
+                        try {
+                            final X500Name x500Name = new JcaX509CertificateHolder(cert).getSubject();
+                            final RDN cn = x500Name.getRDNs(BCStyle.CN)[0];
+                            return IETFUtils.valueToString(cn.getFirst().getValue());
+                        } catch (Exception e) {
+                            logger.warn("Failed to get the common name from a certificate: {}", cert, e);
+                            return null;
+                        }
+                    });
+
+    @Nullable
+    public static String getCommonName(SSLSession session) {
+        final Certificate[] certs = session.getLocalCertificates();
+        if (certs == null || certs.length == 0) {
+            return null;
+        }
+        return getCommonName(certs[0]);
+    }
+
+    @Nullable
+    public static String getCommonName(Certificate certificate) {
+        if (!(certificate instanceof X509Certificate)) {
+            return null;
+        }
+        return commonNameCache.get((X509Certificate) certificate);
+    }
+
+    public static List<X509Certificate> toX509Certificates(File file) throws CertificateException {
+        requireNonNull(file, "file");
+        return ImmutableList.copyOf(SslContextProtectedAccessHack.toX509CertificateList(file));
+    }
+
+    public static List<X509Certificate> toX509Certificates(InputStream in) throws CertificateException {
+        requireNonNull(in, "in");
+        return ImmutableList.copyOf(SslContextProtectedAccessHack.toX509CertificateList(in));
+    }
+
+    private static final class SslContextProtectedAccessHack extends SslContext {
```

Comment:
It seems like this class isn't necessary?

Reply:
The visibility of `SslContext.toX509Certificates(file)` is `protect` so we can access the method directly.

Reply:
Thanks! I understood it. 😉